gdnative-core/src/core_types/vector3.rs

```diff
@@ -10,64 +21,441 @@ pub enum Axis {
+        )
+    }
+
+    /// Returns a diagonal matrix with the vector as main diagonal.
+    ///
+    /// This is equivalent to a Basis with no rotation or shearing and this vector's components set
+    /// as the scale.
+    #[inline]
+    pub fn to_diagonal_matrix(self) -> Basis {
+        Basis::from_diagonal(self)
+    }
 
     #[doc(hidden)]
     #[allow(clippy::wrong_self_convention)]
-    fn to_sys(self) -> sys::godot_vector3;
+    #[inline]
+    pub fn to_sys(self) -> sys::godot_vector3 {
+        unsafe { std::mem::transmute(self) }
+    }
+
     /// Internal API for converting to `sys` representation. Makes it possible to remove
     /// `transmute`s elsewhere.
     #[doc(hidden)]
-    fn sys(&self) -> *const sys::godot_vector3;
+    #[inline]
+    pub fn sys(&self) -> *const sys::godot_vector3 {
+        self as *const _ as *const _
+    }
+
     /// Internal API for converting from `sys` representation. Makes it possible to remove
     /// `transmute`s elsewhere.
     #[doc(hidden)]
-    fn from_sys(v: sys::godot_vector3) -> Self;
-}
-
-impl Vector3Godot for Vector3 {
     #[inline]
-    fn direction_to(self, other: Vector3) -> Vector3 {
-        (other - self).normalize()
+    pub fn from_sys(v: sys::godot_vector3) -> Self {
+        unsafe { std::mem::transmute(v) }
     }
 
     #[inline]
-    fn distance_to(self, other: Vector3) -> f32 {
-        (other - self).length()
+    fn glam(self) -> Vec3A {
+        Vec3A::new(self.x, self.y, self.z)
     }
 
     #[inline]
-    fn distance_squared_to(self, other: Vector3) -> f32 {
-        (other - self).square_length()
+    fn gd(from: Vec3A) -> Self {
+        Self::new(from.x, from.y, from.z)
     }
+}
 
-    #[inline]
-    fn to_sys(self) -> sys::godot_vector3 {
-        unsafe { std::mem::transmute(self) }
+impl AsRef<[f32; 3]> for Vector3 {
+    fn as_ref(&self) -> &[f32; 3] {
+        // SAFETY: Vector3 is repr(C)
+        unsafe { &*(self as *const Vector3 as *const [f32; 3]) }
     }
+}
 
-    #[inline]
-    fn sys(&self) -> *const sys::godot_vector3 {
-        self as *const _ as *const _
-    }
+macro_rules! derive_op_impl {
+    ($trait:ident, $func:ident) => {
+        impl $trait for Vector3 {
+            type Output = Self;
+
+            #[inline]
+            fn $func(self, with: Self) -> Self {
+                Self::gd(self.glam().$func(with.glam()))
+            }
+        }
+    };
+    ($trait:ident, $func:ident, $in_type:ty) => {
+        impl $trait<$in_type> for Vector3 {
+            type Output = Self;
+
+            #[inline]
+            fn $func(self, with: $in_type) -> Self {
+                Self::gd(self.glam().$func(with))
+            }
+        }
+    };
+}
+
+macro_rules! derive_op_impl_rev {
+    ($trait:ident, $func:ident, $in_type:ty) => {
+        impl $trait<Vector3> for $in_type {
+            type Output = Vector3;
+
+            #[inline]
+            fn $func(self, with: Self::Output) -> Self::Output {
+                with * self
```

Comment:
You probably want to change this to `$trait::$func(with, self)`? You're also using `*` for `Div` in the current code.